src/operator/tensor/indexing_op.cu

```diff
@@ -545,6 +545,247 @@ void TakeOpForward<gpu>(const nnvm::NodeAttrs& attrs,
+  IType mean;
+  while (lower_bound < upper_bound) {
+    mean = (lower_bound + upper_bound) / 2;
+    if (id <= sorted_data[mean])
+      upper_bound = mean;
+    else
+      lower_bound = mean + 1;
+  }
+  bool found_row = (sorted_data[lower_bound] == id);
+
+  if (id < vocab_dim) {
+    bounds[id] = (found_row) ? lower_bound : -1;
+  }
+
+  // Binary search to find upper bound: stored at bounds[vocab_dim..2*vocab_dim-1]
+  lower_bound = 0;
+  upper_bound = data_dim - 1;
+  while (lower_bound < upper_bound) {
+    mean = (lower_bound + upper_bound + 1) / 2;
+    if (id >= sorted_data[mean])
+      lower_bound = mean;
+    else
+      upper_bound = mean - 1;
+  }
+  found_row = (sorted_data[upper_bound] == id);
+
+  if (id < vocab_dim) {
+    bounds[vocab_dim + id] = (found_row) ? upper_bound : -1;
+  }
+}
+
+/*
+ * \brief kernel to compute gradient of EmbeddingOp
+ * \param grad_in input gradient data
+ * \param original_index reference to the position at original input data for each index
+ * \param index_bounds lower and upper-bounds positions of each unique index
+ * \param grad_out output gradient data
+ * \param embbedding_dim dimension of the dense embedding
+ * \param vocab_dim maximum number of unique indices in the data array: tokens vocabulary size
+ * \param rows_per_block number of grad_in rows to be computed by each block
+ * \param req write/add/null
+ */
+template <typename LType, typename DType, typename IType>
+__global__ void EmbeddingGradKernel(DType *grad_in,
+                                      const IType *original_index,
+                                      const IType *index_bounds,
+                                      const DType *grad_out,
+                                      const index_t embbedding_dim,
+                                      const index_t vocab_dim,
+                                      const int rows_per_block,
+                                      const int req) {
+  extern __shared__ int sharedmem[];
+  LType* grad_in_row =  reinterpret_cast<LType *>(sharedmem);
+
+  // LType has to be bigger than DType, guarded in the launcher code
+  const int n_val = sizeof(DType) < sizeof(LType) ? sizeof(LType) / sizeof(DType) : 1;
+  const LType *aligned_grad_out = reinterpret_cast<const LType *>(grad_out);
+  LType *aligned_grad_in = reinterpret_cast<LType *>(grad_in);
+  const index_t aligned_emb_dim = embbedding_dim / n_val;
+  DType *my_grad_in_row = reinterpret_cast<DType *>(&grad_in_row[threadIdx.x]);
+  LType Lvalue[1];
+  DType* Dvalues = reinterpret_cast<DType*>(Lvalue);
+
+  for (index_t row=0; row < rows_per_block; ++row) {
+    IType my_row = blockIdx.x * rows_per_block + row;
+    if (my_row < vocab_dim) {
+      // Read lower and upper bounds for current row
+      IType lower_bound = index_bounds[my_row];
+      IType upper_bound = index_bounds[vocab_dim + my_row];
+      int nOccurrences = (lower_bound != -1) ? (upper_bound - lower_bound + 1) : 0;
```

Comment:
Super small hack and probably not worth it, but if you set `upper_bound` to -2 if you did not find the lower bound, then you would not need this conditional here.